### PR TITLE
Fixes processing deployment outputs #1316 #1315

### LINKED ISSUE #1315: Expand reference for resources in template
Currently a placeholder mock is returned whenever the `reference()` function was used.

Since the resource is probably not deployed it is not possible cases to return the resource.

This is problematic is many scenarios as the output may need to be processed in a specific way, and more common in Bicep code as outputs are easily referenced.

For cases when the `reference()` is to a resource included in the template the object can be returned.

### LINKED ISSUE #1316: Outputs are not processed
**Description of the issue**

Deployment outputs are not processed. This is problematic because Bicep code will often reference outputs of a module as inputs to another deployment.

**Expected behaviour**

Resolve outputs when they are referenced in code.

**Module in use and version:**

- Module: PSRule.Rules.Azure
- Version: **1.13.4**

Captured output from `$PSVersionTable`:

```text
Name                           Value
----                           -----
PSVersion                      7.2.2
PSEdition                      Core
GitCommitId                    7.2.2
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by BernieWhite**
## PR Summary

- Added support for referencing resources in template. #1315
  - The `reference()` function can be used to reference resources in template.
  - A placeholder value is still used for resources outside of the template.
- Fixed processing of deployment outputs. #1316

Fixes #1315 
Fixes #1316 

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Other code changes**
  - [x] Unit tests created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: src/PSRule.Rules.Azure/Data/Template/Functions.cs [hunk 1]
```diff
@@ -760,8 +760,25 @@ internal static object Reference(ITemplateContext context, object[] args)
             if (argCount < 1 || argCount > 3)
                 throw ArgumentsOutOfRange(nameof(Reference), args);
 
-            ExpressionHelpers.TryString(args[0], out var resourceType);
-            return new MockResource(resourceType);
+            if (!ExpressionHelpers.TryString(args[0], out var resourceId))
+                throw ArgumentFormatInvalid(nameof(Reference));
+
+            var full = argCount == 3 && ExpressionHelpers.TryString(args[2], out var fullValue) && string.Equals(fullValue, "Full", StringComparison.OrdinalIgnoreCase);
```
Comment (ArmaanMcleod) [inline src/PSRule.Rules.Azure/Data/Template/Functions.cs:766]: Can probably put `"Full"` as a const.

File: src/PSRule.Rules.Azure/Data/Template/Functions.cs [hunk 2]
```diff
@@ -760,8 +760,25 @@ internal static object Reference(ITemplateContext context, object[] args)
...
+                ? GetReferenceResult(resourceValue, full)
+                : new MockResource(resourceId);
+        }
+
+        private static object GetReferenceResult(IResourceValue resource, bool full)
+        {
+            if (resource is DeploymentValue deployment)
+                return full ? deployment : deployment.Properties;
+
+            return full ? resource.Value : resource.Value["properties"].Value<JObject>();
```
Comment (ArmaanMcleod) [inline src/PSRule.Rules.Azure/Data/Template/Functions.cs:781]: Same as above with `"properties"`.

File: src/PSRule.Rules.Azure/Data/Template/ResourceValue.cs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        private sealed class DeploymentProperties : ILazyObject
+        {
+            private readonly DeploymentValue _Deployment;
+            private readonly JObject _Properties;
+            private readonly DeploymentOutputs _Outputs;
+
+            public DeploymentProperties(DeploymentValue deployment)
+            {
+                _Deployment = deployment;
+                _Deployment.Value.TryGetProperty("properties", out _Properties);
```
Comment (ArmaanMcleod) [inline src/PSRule.Rules.Azure/Data/Template/ResourceValue.cs:92]: Same as above.

File: src/PSRule.Rules.Azure/Data/Template/ResourceValue.cs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+        public void AddOutput(string name, ILazyValue output)
+        {
+            _Outputs.Add(name, output);
+        }
+
+        public bool TryProperty(string propertyName, out object value)
+        {
+            value = null;
+            if (StringComparer.OrdinalIgnoreCase.Equals(propertyName, "properties"))
```
Comment (ArmaanMcleod) [inline src/PSRule.Rules.Azure/Data/Template/ResourceValue.cs:151]: Same as above.

**Review by ArmaanMcleod — APPROVED** [review]
@BernieWhite Looks great. Just added a few minor suggestions.

**Comment by BernieWhite** [comment]
@ArmaanMcleod Good suggestions.

**Pull request merged**